clang/lib/CIR/Dialect/IR/CIRDialect.cpp

```diff
@@ -513,7 +513,52 @@ LogicalResult CastOp::verify() {
   llvm_unreachable("Unknown CastOp kind?");
 }
 
+Value fold_casts(llvm::SmallVector<CastOp>& ops) {
+  if (ops.size() < 2)
+    return {};
+
+  auto head = *ops.begin();
+  auto tail = *ops.rbegin();
+
+  // if bool_to_int -> ...  -> int_to_bool: take the bool
+  // as we had it was before all casts
+  if (head.getKind() == mlir::cir::CastKind::bool_to_int
+      && tail.getKind() == mlir::cir::CastKind::int_to_bool)
+    return head.getSrc();
+
+  // if int_to_bool -> ...  -> int_to_bool: take the result
+  // of the first one, as no other casts (and ext casts as well)
+  // don't change the first result
+  if (head.getKind() == mlir::cir::CastKind::int_to_bool
+      && tail.getKind() == mlir::cir::CastKind::int_to_bool)
+    return head.getResult();
+
+  return {};
+}
+
+bool is_int_or_bool_cast(mlir::cir::CastOp op) {
+  auto kind = op.getKind();
+  return kind == mlir::cir::CastKind::bool_to_int ||
+          kind == mlir::cir::CastKind::int_to_bool ||
+          kind == mlir::cir::CastKind::integral;
+}
+
+void collect_casts(CastOp op, llvm::SmallVector<CastOp>& ops) {
+  if (!is_int_or_bool_cast(op))
+    return;
+
+  if (auto par = dyn_cast_or_null<CastOp>(op.getSrc().getDefiningOp()))
+    collect_casts(par, ops);
+
+  ops.push_back(op);
+}
+
 OpFoldResult CastOp::fold(FoldAdaptor adaptor) {
+  llvm::SmallVector<CastOp> casts;
+  collect_casts(*this, casts);
+  if (auto v = fold_casts(casts))
+    return v;
```

Comment:
This should be unified with the switch below, is it really necessary to collect them? I feel if you handle each below the pass will run again until everything is covered, no need for us to do the extra work here.

Reply:
Let's say I have the next casts:
```
%1 = cir.cast(bool_to_int, %0 : !cir.bool), !s32i
%2 = cir.cast(integral, %1 : !s32i), !s64i
%3 = cir.cast(int_to_bool, %2 : !s64i), !cir.bool
```
The desired result is to return `%0` from `fold` for the cast `%3`, right? In this case `%3` will be removed, the number of users for `%2` will be equal to zero and the operation will be removed also, and the same will be done for `%1`, so all the casts will be eliminated.  
Now, if I try to do it just using folds, the only thing I could use is the result of `fold` for `%2` , but there is no evidence I can return some value when processing it.

So looks like I need to collect casts somehow. Is it really  possible via `fold` ? If it is, I will do it! But so far I don't see how to do it, sorry( 
Any thoughts?